doctrine/event_listeners_subscribers.rst

```diff
@@ -7,18 +7,122 @@
-objects whenever a certain action (e.g. ``prePersist()``) happens within Doctrine.
-This could be useful, for example, to create an independent search index
-whenever an object in your database is saved.
-
-Doctrine defines two types of objects that can listen to Doctrine events:
-listeners and subscribers. Both are very similar, but listeners are a bit
-more straightforward. For more, see `The Event System`_ on Doctrine's website.
+Doctrine has a rich event system that fires when almost anything happens
+inside the system. This means that you can create arbitrary
+:doc:`services </service_container>` and configure Doctrine to call those
+whenever a certain action occurs (e.g. an entity gets persisted).
+So instead of having to call some service *every single time* whenever
+you're doing something in the database ...
+
+    $userService->doWhateverPrePersist();
+    $em->persist($user);
+
+... you could easily set up `doWhateverPrePersist()` as a listener and
+then just call `$em->persist($user);` in your main code.
+
+As a first step, you should decide which event you want to listen for. Here
+are some popular events - for a full list see `Doctrine's Lifecycle Events`_:
+
+* `prePersist` and `postPersist`
+* `preUpdate` and `postUpdate`
+* `preRemove` and `postRemove`
+* `preFlush` and `postFlush`
+
+The second question would be:
+
+* Should your service only listen for a *specific* entity? Then go for
+  Entity Listeners
+* Or should your service be called for *all* entities (or at least
+  *several* entities)? Then go for Event Listeners or Event Subscribers
+  
+Entity Listeners
+----------------
+
+To set up an entity listener, start by creating a listener class:
+
+.. code-block:: php
+
+    // src/EventListener/UserListener
+    namespace App\EventListener;
+
+    use Doctrine\ORM\Event\LifecycleEventArgs;
+    use App\Entity\User;
+
+    class UserListener
+    {    
+        public function prePersist(User $user, LifecycleEventArgs $event)
+        {
+            // ...
+        }
+    }
+
+Next, you have to register the class as a listener. There are two ways to
+achieve this:
+
+You can either do the registration in your entity like this:
+
+.. code-block:: php
+
+    // src/Entity/User.php
+    namespace App\Entity\User;
+
+    use Doctrine\ORM\Mapping as ORM;
+
+    /**
+     * @ORM\Entity
+     * @ORM\EntityListeners({"App\EventListener\UserListener"})
+     */
+    class User
+    {
+        // ...
+    }
+
+This works fine, but your event listener will not be registered as a service
+(even if you have ``autowire: true`` in your ``services.yaml``). So to register
+it as a service, you have to add this to your ``services.yaml``:
 
-The Doctrine website also explains all existing events that can be listened to.
+.. configuration-block::
+
+    .. code-block:: yaml
+
+        services:
+            App\EventListener\UserListener:
+                tags:
+                    - { name: doctrine.orm.entity_listener }
```

Comment:
In order to leverage the `autowire: true` feature, you can create and empty interface `EntityListenerInterface` which all your entity listeners implement, then add this to your `services.yaml`:

```
services:
    _instanceof:
        App\EventListener\EntityListenerInterface:
            tags: [doctrine.orm.entity_listener]
```
Then you don't need to define all your entity listeners as services.

Reply:
That's a good point. You've got some merge conflicts at the moment though. Once you fix them up I'll approve :+1:

Reply:
@ThomasLandauer you can build the documentation locally with your changes, https://symfony.com/doc/current/contributing/documentation/overview.html#build-the-documentation-locally that way you can see how it looks before it gets merged